### PR TITLE
OCPQE-18265:Support private endpoint for registry when install a prviate azure cluster

### PR DESCRIPTION
**Pull request by wewang58**
@jianlinliu  @jinyunma  Could you help to review it, thanks

### CODE REVIEW

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-proxy-amd-f28-destructive

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-proxy-arm-f14

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-azure-ipi-fullyprivate-proxy-f14

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-arm64-nightly-azure-ipi-fullyprivate-proxy-f14

**Comment by wewang58** [comment]
cc @flavianmissi ^^

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 1]
```diff
@@ -565,6 +565,7 @@ tests:
       OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE: release:multi-latest
     env:
       BASE_DOMAIN: qe.azure.devcluster.openshift.com
+      CONFIG_PRIVATE_ENDPOINT_FOR_REGISTRY: "true"
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:568]: public registry in a private cluster is still a choice for customers, right? If yes, I think maybe deserve a dedicated installworkflow to cover it.

**Comment by wewang58** [comment]
> This looks great! Do we know if there are any tests that expose the registry's `defaultRoute` and use it? If we do that somewhere then we also want to disable redirect, i.e set `disableRedirect: true` in the registry operator config yaml.

i will add post-check  test in our auto scripts in future, which will not be in the repo, it's in our qe repo.

File: ci-operator/step-registry/ipi/conf/azure/image-registry/ipi-conf-azure-image-registry-ref.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+ref:
+  as: ipi-conf-azure-image-registry
+  from_image:
+    namespace: ocp
+    name: "4.12"
```
[the remark this answers is not in the document]
Comment (wewang58) [reply ci-operator/step-registry/ipi/conf/azure/image-registry/ipi-conf-azure-image-registry-ref.yaml:5]: Yes, i think it's ok @jinyunma can you confirm it, thanks

[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/azure/image-registry/ipi-conf-azure-image-registry-ref.yaml:5]: In the step, it just write some config file, not have to depends any version of oc or installer, so I think it is okay.

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-amd-f14

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-amd-f14

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-amd-f14

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-amd-f14

**Comment by wewang58** [comment]
@jianlinliu  @jinyunma  @flavianmissi  Updated according your comments, and seems job is ok, could you review again, thanks

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 3]
```diff
@@ -571,6 +571,20 @@ tests:
     test:
     - chain: openshift-e2e-test-qe-destructive
     workflow: cucushift-installer-rehearse-azure-ipi-fullyprivate-proxy
+- as: azure-ipi-fullyprivate-internal-registry-amd-f14
```
Comment (jinyunma) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:574]: missing job to cover tests for destructive and arm platform, suggest to add below two jobs:
- azure-ipi-fullyprivate-internal-registry-amd-f28-destrcutive
- azure-ipi-fullyprivate-internal-registry-arm-f14

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-arm-f14

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-amd-f28-destructive

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 4]
```diff
@@ -571,6 +571,34 @@ tests:
...
+- as: azure-ipi-fullyprivate-internal-registry-amd-f28-destructive
+  cron: 44 14 20 * *
+  steps:
+    cluster_profile: azure-qe
+    dependencies:
+      OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE: release:multi-latest
+    env:
+      BASE_DOMAIN: qe.azure.devcluster.openshift.com
+      E2E_RUN_TAGS: '@amd64 and @azure-ipi and @network-ovnkubernetes and not @fips
+        and @proxy'
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:583]:
```suggestion
        '
```

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 5]
```diff
@@ -571,6 +571,34 @@ tests:
     test:
     - chain: openshift-e2e-test-qe-destructive
     workflow: cucushift-installer-rehearse-azure-ipi-fullyprivate-proxy
+- as: azure-ipi-fullyprivate-internal-registry-amd-f28-destructive
+  cron: 44 14 20 * *
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:575]: psl run `ci-operator/config/openshift/openshift-tests-private/tools/generate-cron-entry.sh <job-name> <ci-yaml_file_name>` to generate an uniq cron.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml [hunk 6]
```diff
@@ -587,6 +615,22 @@ tests:
     test:
     - chain: openshift-e2e-test-qe
     workflow: cucushift-installer-rehearse-azure-ipi-fullyprivate-proxy
+- as: azure-ipi-fullyprivate-internal-registry-arm-f14
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:618]: This one should be moved to line 588, and replace it?

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/fullyprivate/internal-registry/OWNERS [hunk 7]
```diff
@@ -0,0 +1,14 @@
+approvers:
+- jianlinliu
+- Amoghrd
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/fullyprivate/internal-registry/OWNERS:3]: I think you can remove this user now.

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/fullyprivate/internal-registry/OWNERS [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+- yunjiang29
+- mgahagan73
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/fullyprivate/internal-registry/OWNERS:5]: I think you can remove this user now, and add @jinyunma to replace it.

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-amd-f28-destructive

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-arm-f14

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml (hunk 4 above)
Comment (wewang58) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml]: @jianlinliu  Updated all according your comments, let's wait job result,thanks

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml (hunk 6 above)
Comment (jianlinliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:603]: We should move the job behind azure-ipi-fullyprivate-internal-registry-amd-f28-destructive to make them in order.

Comment (jianlinliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:603]: and the pair of `azure-ipi-fullyprivate-internal-registry` jobs should be placed before `azure-ipi-fullyprivate-proxy` jobs.

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/fullyprivate/internal-registry/deprovision/cucushift-installer-rehearse-azure-ipi-fullyprivate-internal-registry-deprovision-chain.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+chain:
+  as: cucushift-installer-rehearse-azure-ipi-fullyprivate-internal-registry-deprovision
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/fullyprivate/internal-registry/deprovision/cucushift-installer-rehearse-azure-ipi-fullyprivate-internal-registry-deprovision-chain.yaml:2]: Seem like you can directly reference `cucushift-installer-rehearse-azure-ipi-fullyprivate-deprovision` chain in the install workflow, does not need to introduce this dedicated provision chain.

**Comment by wewang58** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-multi-nightly-azure-ipi-fullyprivate-internal-registry-arm-f14

File: ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/fullyprivate/internal-registry/deprovision/cucushift-installer-rehearse-azure-ipi-fullyprivate-internal-registry-deprovision-chain.yaml (hunk 9 above)
Comment (wewang58) [reply ci-operator/step-registry/cucushift/installer/rehearse/azure/ipi/fullyprivate/internal-registry/deprovision/cucushift-installer-rehearse-azure-ipi-fullyprivate-internal-registry-deprovision-chain.yaml]: Sorry, update late, just computer meets black screen before

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml (hunk 6 above)
Comment (jianlinliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml:603]: @wewang58 pls do not forget this comments.

Comment (wewang58) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__multi-nightly.yaml]: @jianlinliu  Oh, updated

**Comment by jianlinliu** [comment]
cc @liangxia to review.

**Pull request merged**